nntrainer/src/layers.cpp

```diff
@@ -408,87 +342,90 @@ void OutputLayer::copy(Layer *l) {
   this->loss = from->loss;
 }
 
-Tensor OutputLayer::backwarding(Tensor label, int iteration) {
+Tensor FullyConnectedLayer::backwarding(Tensor derivative, int iteration) {
+  Tensor djdb;
   float loss_sum = 0.0;
-  Tensor y2 = label;
+  Tensor y2 = derivative;
   Tensor y;
-  if (activation_type == ACT_SOFTMAX)
-    y = hidden.apply(softmax);
-  else
-    y = hidden.apply(activation);
-
-  Tensor ret;
-  Tensor djdb;
-
-  float ll = opt.getLearningRate();
-  if (opt.getDecaySteps() != -1) {
-    ll = ll * pow(opt.getDecayRate(), (iteration / opt.getDecaySteps()));
-  }
-
-  switch (cost) {
-  case COST_MSR: {
-    Tensor sub = y2.subtract(y);
-    Tensor l = (sub.multiply(sub)).sum().multiply(0.5);
-    std::vector<float> t = l.mat2vec();
-    for (int i = 0; i < l.getBatch(); i++) {
-      loss_sum += t[i];
-    }
-
-    loss = loss_sum / (float)l.getBatch();
-    if (opt.getWeightDecayType() == WeightDecayType::l2norm) {
-      loss += opt.getWeightDecayLambda() * 0.5 * (weight.l2norm());
-    }
-    if (activation_type == ACT_SOFTMAX) {
-      djdb = y.subtract(y2).multiply(y.apply(softmaxPrime));
-    } else {
-      djdb = y.subtract(y2).multiply(hidden.apply(activation_prime));
-    }
-  } break;
-  case COST_ENTROPY: {
-    Tensor l;
-    if (activation_type == ACT_SIGMOID) {
-      djdb = y.subtract(y2).multiply(1.0 / y.getWidth());
-      l = (y2.multiply(y.apply(logFloat))
-             .add((y2.multiply(-1.0).add(1.0))
-                    .multiply((y.multiply(-1.0).add(1.0)).apply(logFloat))))
-            .multiply(-1.0 / (y2.getWidth()))
-            .sum();
-    } else if (activation_type == ACT_SOFTMAX) {
-      djdb = y.subtract(y2).multiply(1.0 / y.getWidth());
-      l =
-        (y2.multiply(y.apply(logFloat))).multiply(-1.0 / (y2.getWidth())).sum();
-    } else {
-      ml_loge("Only support sigmoid & softmax for cross entropy loss");
-      exit(0);
-    }
 
-    std::vector<float> t = l.mat2vec();
-
-    for (int i = 0; i < l.getBatch(); i++) {
-      loss_sum += t[i];
+  if (!last_layer) {
+    djdb = derivative.multiply(hidden.apply(activation_prime));
```

Comment:
Q. Please consider that I am just studying this part. If it is not last layer, is there any need for this?

Reply:
Well... Why not? FC layer can be used in the middle of network.